### PR TITLE
Trainer with metrics

### PR DESCRIPTION
**Pull request by vfdev-5**
A proposition to update `create_supervised_trainer` to accept metrics similarly to `create_supervised_evaluator`. A log can be updated as the following
```
Epoch[1] Iteration[910/938] Loss: 0.48
Epoch[1] Iteration[920/938] Loss: 0.56
Epoch[1] Iteration[930/938] Loss: 0.34
Training Results - Epoch: 1  Avg accuracy: 0.67 Avg loss: 0.97
Validation Results - Epoch: 1  Avg accuracy: 0.94 Avg loss: 0.21
```
What do you guys think ?

In addition, added minor changes:
- Remove `Variable`
- replace tensorboardx graph creation method
- fix `F.log_softmax` warnings on unspecified dimension

### CODE REVIEW

**Comment by vfdev-5** [comment]
Travis CI is need to be restarted

**Comment by alykhantejani** [comment]
@vfdev-5 done

**Comment by vfdev-5** [comment]
Thank you @alykhantejani

File: docs/source/quickstart.rst [hunk 1]
```diff
@@ -17,7 +17,11 @@ Code
     optimizer = torch.optim.SGD(model.parameters(), lr=0.01, momentum=0.8))
     loss = torch.nn.NLLLoss()
 
-    trainer = create_supervised_trainer(model, optimizer, loss)
+    trainer = create_supervised_trainer(model, optimizer, criterion,
```
Comment (jasonkriss) [inline docs/source/quickstart.rst]: Small typo here... `criterion` needs to be `loss` or vice versa

Comment (vfdev-5) [reply docs/source/quickstart.rst]: thanks @jasonkriss !

**Comment by jasonkriss** [comment]
Thanks for taking this on @vfdev-5.

I'm still a lil hesitant on this though. I get that this is a pretty common pattern so it might makes sense to add the ability to `create_supervised_trainer` but i hesitate to put it in the quickstart as I would rather not encourage this approach.

In your lil output sample, the training results are presented as if they are comparable. However, as training metrics are computed against a moving target, they are not. I would much rather encourage users to use the evaluator on both a train set and a validation set.

For example, this is (roughly) what I have in every project:

```python
train_dataset = ...
validation_dataset = ...
train_eval_dataset = Subset(train_dataset, list(range(len(validation_dataset))))
train_loader = DataLoader(train_dataset, ...)
train_eval_loader = DataLoader(train_eval_dataset, ...)
validation_loader = DataLoader(validation_dataset, ...)

trainer = create_supervised_trainer(...)
evaluator = create_supervised_evaluator(...)

@trainer.on(Events.EPOCH_COMPLETED)
def evaluate(trainer):
  validation_metrics = evaluator.run(validation_loader).metrics
  train_metrics = evaluator.run(train_eval_loader).metrics
  print('validation', validation_metrics)
  print('train', train_metrics)
```

That way, we can actually meaningfully compare training and validation metrics.

What do you all think?

**Comment by vfdev-5** [comment]
@jasonkriss probably I see what you mean. If I understand correctly your snippet than `train_eval_dataset` is not seen by model for training, I suppose. So, then your metrics show model's generalization capacity if two dataset are from different distributions?
In my case, train vs validation metrics can be used to see the overfitting.
Correct me please if I missed something

**Comment by vfdev-5** [comment]
@jasonkriss I read  again the snippet, actually, I think I am wrong saying about `train_eval_dataset` is not seen by model for training. Then, some questions:
- why do you prefer to have the same size `len(train_eval_dataset) == len(validation_dataset)`, train_eval_dataset as validation dataset and not taking this randomly ?

I agree that computing train metrics during the training is a sort of average on something inexistant :)

**Comment by vfdev-5** [comment]
@jasonkriss thank you for explaination ! 
I agree that computing metrics with `create_supervised_trainer` does not have much sense, so no need for such modification.
Okay, I'll see what can be kept here in this PR...

**Comment by jasonkriss** [comment]
👍 sounds good. Thanks @vfdev-5. @alykhantejani might have some thoughts on this as well. I know we've discussed it a bit in the past.

**Comment by vfdev-5** [comment]
Ah, okay, surely I missed this discussion.
Maybe, I could reimplement your pattern of train metrics computation in `mnist*.py` examples and tell about this in `quickstart.rst`. 
Otherwise I think I wont be the last person who would like to propose such PR then :)

**Comment by vfdev-5** [comment]
@jasonkriss in your practice do you reuse the same `evaluator` for training and validation performance evaluation ? In some sense, it is better to have two evaluators because validation evaluator can have handlers like ModelCheckpoint, EarlyStopping and reduce on plateau. 
So, it seems to be tricky to verify which (train or val) metrics are in the state in the handler...

**Comment by alykhantejani** [comment]
Sorry I haven't had a chance to look at this yet (have been away for a few days), I will get to this first thing tomorrow when I'm back :)

**Comment by alykhantejani** [comment]
@vfdev-5 the discussion on this previously can be found [here](https://github.com/pytorch/ignite/issues/118#issuecomment-374296777). The conclusion we came to was that:
1) it is indeed useful to compute metrics on the training set. 
2) sometimes you don't want to have a separate evaluator run over your training set (or a subset of it) just to compute metrics when these can be done whilst iterating over the training set during trianing
3) with this in mind a "trainer" and an "evaluator" are essentially the same thing, i.e. an "engine"

After that merge to have a single engine, metrics can now be added to the Trainer no problem. However, as @jasonkriss  mentions this is probably not the best idea as the underlying model moves every iteration.  

So maybe it's worthwhile mentioning in one of the readme docs that whilst this is possible (as you can attach metrics to any engine), the recommended approach is to have a separate engine that is used to benchmark the model on the training set (or a subset) at regular intervals. Therefore we should not add this functionality to `create_supervised_trainer` as we don't want to encourage this pattern.

**Comment by vfdev-5** [comment]
@jasonkriss @alykhantejani I removed `create_supervised_trainer` with metrics and left minor updates of examples, reimplemented metrics computation with 2nd pass and added a footnote on these two approaches.

File: examples/mnist_with_visdom.py [hunk 2]
```diff
@@ -34,7 +34,7 @@ def forward(self, x):
         x = F.relu(self.fc1(x))
         x = F.dropout(x, training=self.training)
         x = self.fc2(x)
-        return F.log_softmax(x)
+        return F.log_softmax(x, dim=-1)
```
Comment (jasonkriss) [inline examples/mnist_with_visdom.py]: Is there a reason you didn't add a handler to log validation results in this example?

**Comment by jasonkriss** [comment]
@vfdev-5 just one small comment/question. Looks great! Thanks!

File: examples/mnist_with_visdom.py (hunk 2 above)
Comment (vfdev-5) [reply examples/mnist_with_visdom.py]: @jasonkriss do you mean no `def log_training_results(engine)` as in mnist or mnist with tensorboardX ? 
Actually, there is no particular reason except we need to create specific windows and draw lines as done in validation... :)
Yep, I will add this

Comment (vfdev-5) [reply examples/mnist_with_visdom.py]: @jasonkriss I updated the example, if you want to take a look

Comment (jasonkriss) [reply examples/mnist_with_visdom.py]: @vfdev-5 LGTM. Thanks!
